### PR TITLE
3/n integrate new LightningDataFetcher into loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #8891 
Integrates the fetcher into the training loop. Originally introducded in #8890 (with tests)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/profiler/base.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def profile(self, action_name: str) -> None:
         finally:
             self.stop(action_name)
 
-    def profile_iterable(self, iterable, action_name: str) -> None:
+    def profile_iterable(self, iterable, action_name: str) -> Generator:
```
Comment (ananthsub) [inline pytorch_lightning/profiler/base.py:99]:
```suggestion
    def profile_iterable(self, iterable: Iterable, action_name: str) -> Generator:
```
?

Comment (tchaton) [reply pytorch_lightning/profiler/base.py:99]: Done !

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -98,7 +99,9 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
         """Performs evaluation on one single dataloader"""
         void(*args, **kwargs)
         dataloader = self.trainer.accelerator.process_dataloader(self.current_dataloader)
-        dataloader_iter = enumerate(dataloader)
+        data_fetcher = DataFetcher()
```
Comment (justusschock) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:102]: the logic to change this to a custom class is yet to come?

File: pytorch_lightning/profiler/base.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def profile(self, action_name: str) -> None:
         finally:
             self.stop(action_name)
 
-    def profile_iterable(self, iterable, action_name: str) -> None:
+    def profile_iterable(self, iterable: Iterable, action_name: str) -> Generator:
```
Comment (justusschock) [inline pytorch_lightning/profiler/base.py:99]: can we also add typing for the generator? i.e. what does it yield?

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 4]
```diff
@@ -60,9 +61,10 @@ def on_trainer_init(
         self.trainer._is_data_prepared = False
 
     def get_profiled_train_dataloader(self, train_dataloader):
-        profiled_dl = self.trainer.profiler.profile_iterable(
-            enumerate(prefetch_iterator(train_dataloader)), "get_train_batch"
-        )
+        self.prefetcher = DataFetcher()
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:64]: why isn't this done inside the loop (similar to evaluation loop)?

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 5]
```diff
@@ -26,6 +26,7 @@ class DataConnector:
     def __init__(self, trainer: "pl.Trainer", multiple_trainloader_mode: str = "max_size_cycle"):
         self.trainer = trainer
         self.multiple_trainloader_mode = multiple_trainloader_mode
+        self.prefetcher: Optional[DataFetcher] = None
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:29]: how is this one considered if passed?

**Review by ananyahjha93 — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:102]: Yes.

Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:102]: For now, this is just used internally.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:64]: The next PR will add profiling, and I will move the logic from the evaluation loop to the data_connector too.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:29]: Not sure to fully grasp the question.

File: pytorch_lightning/profiler/base.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/profiler/base.py:99]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py]: But we will address this when the `data_fetcher` is made public and can be passed through the Trainer arguments.
